### PR TITLE
Create reuseport v2 using Go 1.11

### PR DESCRIPTION
**Pull request by Kubuxu**
Should wait for Go 1.11 is released.
578 lines of this (95%) of this are test cases.

### CODE REVIEW

File: v2/reuse_test.go [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+type readOnly struct {
+	r io.Reader
+}
+
+func (r readOnly) Read(buf []byte) (n int, err error) {
+	return r.r.Read(buf)
+}
+
+func echo(c net.Conn) {
+	io.Copy(c, readOnly{c})
```
Comment (Kubuxu) [inline v2/reuse_test.go]: This is to remove optimization added in Go 1.11 which uses splice but also doubles required file descriptors.

The optimization is only used when copying from TCP to TCP.

**Comment by Kubuxu** [comment]
@Stebalien do we have to set linger ? AFAIK it can be set using normal Go functions.
Also the nodelay is set to `true` by default in Go.

**Comment by Stebalien** [comment]
You're right, we shouldn't have to set NoDelay. However, I think we still have to set linger on *listen*, right? I'm not sure. IIRC, I ran into some bug without setting these but now I can't remember what. Feel free to revert that commit.

**Comment by Kubuxu** [comment]
We can set linger in upstream if it is needed. Then we can use Golang lib functions so the compability problems don't baloon on us.

**Comment by Stebalien** [comment]
@Kubuxu do you have time to finish this (i.e., factor everything into their proper places)?

**Comment by Kubuxu** [comment]
If go-libp2p isn't dependency mess at this time, I can try doing this tomorrow.

This should be good to merge as it is defined in a separate namespace (yey for the idea of semantic import versioning).

**Comment by Stebalien** [comment]
> This should be good to merge as it is defined in a separate namespace (yey for the idea of semantic import versioning).

How? We don't *currently* use go.mod files. Do you mean just merge it into a different directory?

And no, go-libp2p is in a pretty solid state ATM.

**Comment by Kubuxu** [comment]
Yes, the merge currently is to `v2` directory so old version will still function. Not so important in case of this package (there aren't many users of it).

**Comment by Stebalien** [comment]
What's the state of this?

**Comment by anacrolix** [comment]
I'm going to look into this.

**Comment by Stebalien** [comment]
Superseded by #56.

**Pull request closed without merging**